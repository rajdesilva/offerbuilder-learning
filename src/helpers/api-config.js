import { init as initApm } from "@elastic/apm-rum";

export const baseURL = {
  API_URL_BASE: process.env.REACT_APP_BASE_API_URL,
  LOGIN_URL: process.env.REACT_APP_LOGIN_URL,
  LOGOUT_URL: process.env.REACT_APP_LOGOUT_URL,
  PROFILE_URL: process.env.REACT_APP_USER_PROFILE_URL,
  HISTORY_URL: process.env.REACT_APP_USER_HISTORY_URL,
};

export const apm = initApm({
  // Set required service name (allowed characters: a-z, A-Z, 0-9, -, _, and space)
  serviceName: process.env.REACT_APP_APM_SERVICE_NAME,
  serverUrl: process.env.REACT_APP_APM_SERVICE_URL,

  // Set service version (required for sourcemap feature)
  serviceVersion: process.env.REACT_APP_APM_SERVICE_VERSION,
});