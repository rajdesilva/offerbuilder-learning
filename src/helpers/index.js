export * from './api-config';
export * from './history';
