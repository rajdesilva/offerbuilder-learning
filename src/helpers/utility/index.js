export { createSearchQueryURL } from './createSearchQueryURL';
export { isParentOrChildrenPropertyAdded } from './isParentOrChildrenPropertyAdded';
export { isPropertySelected } from './isPropertySelected';
export { mapBrands } from './mapBrands';
export { mapOnlySuppliers } from './mapOnlySuppliers';
export { parseFormDescriptionsAndImagesForOffer } from './parseFormDescriptionsAndImagesForOffer';
export { removePropertyForSelectedChannels } from './removePropertyForSelectedChannels';
export { removePropertyForSelectedSuppliers } from './removePropertyForSelectedSuppliers';
export { getParsedProperty } from './getParsedProperty';
export { getFormattedValue } from './getFormattedValue';
export { getValue } from './getValue';
export { parsePropertyMarketingInfoToFormData } from './parsePropertyMarketingInfoToFormData';
export { createOfferRequestObject } from './createOfferRequestObject';
export { getParsedOffers } from './getParsedOffers';
export { getListOfPropertiesForSupplier } from './getListOfPropertiesForSupplier';
export { getListOfPropertiesForChannels } from './getListOfPropertiesForChannels';
export { getInitializedState } from './getInitializedState';
export { getParsedNewOfferSettingsObject } from './getParsedNewOfferSettingsObject';
export { getListOfPropertiesForStorefront } from './getListOfPropertiesForStorefront';
export { checkIfMultipleChannelPresentForTarget } from './checkIfMultipleChannelPresentForTarget';
export { checkIfMultipleSupplierPresentForTarget } from './checkIfMultipleSupplierPresentForTarget';
export { checkIfMultipleStorefrontPresentForTarget } from './checkIfMultipleStorefrontPresentForTarget';
export { removePropertyForSelectedStorefront } from './removePropertyForSelectedStorefront';
export { checkIfMultipleBrandPresentForTarget } from './checkIfMultipleBrandPresentForTarget';
export { getListOfPropertiesForBrand } from './getListOfPropertiesForBrand';
export { getSearchStateFromRedux } from './getSearchStateFromRedux';
export { getObjectStringify } from './getObjectStringify';
export { parseOfferMarketingFromOffer } from './parseOfferMarketingFromOffer';
export { parseOfferMarketingLanguagesFromOffer } from './parseOfferMarketingLanguagesFromOffer';
export { getFilterData } from './getFilterData';
export { getSearchAndFilterQueryURL } from './getSearchAndFilterQueryURL';
export { getSearchQueryURLForEditAddProperties } from "./getSearchQueryURLForEditAddProperties";
export { isSavedOfferProperty } from './isSavedOfferProperty';
export { getParsedPropertiesForDeeplinkSettings } from "./getParsedPropertiesForDeeplinkSettings";
export { resetApplicationReduxAndStorage } from "./resetApplicationReduxAndStorage";
export { updateSettingsBrandWithSearchedValues } from "./updateSettingsBrandWithSearchedValues";
export { checkIfUserHasRole } from "./checkIfUserHasRole";
export { getUserOfferBuilderRole } from "./getUserOfferBuilderRole";
export { checkIfNewPropertiesAdded } from "./checkIfNewPropertiesAdded";
export { getAllBrandsWithoutKeysFromTempBrands } from "./getAllBrandsWithoutKeysFromTempBrands";
export { checkIfItemAlreadyPresentInList } from "./checkIfItemAlreadyPresentInList";
export { checkIfTravellingDateValidAsPerBookingDate } from "./checkIfTravellingDateValidAsPerBookingDate";
export { checkIfPropertiesHaveMultipleLocations } from "./checkIfPropertiesHaveMultipleLocations";
export { displayMargin } from "./displayMargin";
export { getFilterCount } from "./getFilterCount";
export { labelForBrands } from "./labelForBrands";
export { getDistanceFromLatLonInKm } from "./getDistanceFromLatLonInKm";
export { getTrustYouIdListFromProperties } from "./getTrustYouIdListFromProperties";
export { getSearchParamsToFetchMarketPrice } from "./getSearchParamsToFetchMarketPrice";
export { getMarketPriceForPropertyFromList } from "./getMarketPriceForPropertyFromList";
export { getImageFilterCount } from "./getImageFilterCount";
export { getSearchAndFilterImagesQueryUrl } from "./getSearchAndFilterImagesQueryUrl";
export { getImageName } from "./getImageName";
export { updateReduxWithOfferData } from "./updateReduxWithOfferData";
export { getOfferTypeFromPropertiesSelected } from "./getOfferTypeFromPropertiesSelected";