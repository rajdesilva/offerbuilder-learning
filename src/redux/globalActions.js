export const globalActions = {
  'SET_SUCCESS': 'SET_SUCCESS',
  'SET_ERROR': 'SET_ERROR',
  'SET_LANGUAGE': 'SET_LANGUAGE',
  'CLEAR': 'CLEAR',
}