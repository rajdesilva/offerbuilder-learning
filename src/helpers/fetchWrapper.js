import { message } from "antd";
import { createIntl, createIntlCache } from "@formatjs/intl";
import * as languages from "../langs/";
import { baseURL } from "./api-config";
import { history } from "./history";

// This is optional but highly recommended
// since it prevents memory leak
const cache = createIntlCache();

// Create the `intl` object
const intl = createIntl(
  {
    // Locale of the application
    // TODO:
    locale: "en",
    // Locale of the fallback defaultMessage
    defaultLocale: "en",
    messages: languages["en"],
  },
  cache
);
const defaultRequest = {
  credentials: "include",
  headers: {
    "Content-Type": "application/json",
  },
};
export const fetchWrapper = (url, payload) => {
  payload = {
    ...defaultRequest,
    ...payload,
  };
  return fetch(url, payload).then(
    (response) => {
      // Check that the response is valid and reject an error
      // response to prevent subsequent attempt to parse json
      if (!response.ok) {
        if (response.status === 500) {
          message.error(
            intl.formatMessage({
              id: "nemo.error500msg",
            })
          );
        } else if (response.status === 503) {
          message.error(
            intl.formatMessage({
              id: "nemo.error503msg",
            })
          );
        } else if (response.status === 504) {
          message.error(
            intl.formatMessage({
              id: "nemo.error504msg",
            })
          );
        } else {
          message.error(
            intl.formatMessage({
              id: "nemo.somethingWentWrongPleaseTryAgain",
            })
          );
        }
        return;
      }
      if (response.status === 403) {
        history.push("/access-denied");
      }
      return response.json ? response.json() : response;
    },
    (error) => {
      /**
       * user url, payload and error here for sending to the logger.
       */
      if (error instanceof TypeError) {
        window.location.href = baseURL.API_URL_BASE + "redirect";
      }
      console.error("fetch error : ", error);
    }
  );
};
