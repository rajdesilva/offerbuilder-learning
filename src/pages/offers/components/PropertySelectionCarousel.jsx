import React, { Fragment } from "react";
import ImageGallery from "react-image-gallery";
import "react-image-gallery/styles/css/image-gallery.css";
import { Button, Image, Row, Checkbox, Modal } from "antd";
import { FormattedMessage } from "react-intl";
import {
  CloseOutlined,
  LeftCircleFilled,
  RightCircleFilled,
} from "@ant-design/icons";
import defaultImg from "../../../assets/images/default-img.svg";
import { getImageName } from "../../../helpers/utility";

export default function PropertySelectionCarousel(props) {
  let currentImage = "";

  const onChange = (checkedValues) => {
    if (
      props.selectedImages.findIndex(
        (image) => image === checkedValues.target.value
      ) === -1 &&
      checkedValues.target.checked
    ) {
      // true to add
      props.updateSelectedImages(currentImage, true);
    } else {
      // false to remove
      props.updateSelectedImages(currentImage, false);
    }
  };

  const imgList =
    props.images &&
    props.images.map((image, index) => {
      return {
        original: image,
        key: index,
      };
    });

  if (imgList && imgList.length === 0) {
    props.hideCarousel();
  }
  return (
    <Modal
      visible={true}
      closable={true}
      onCancel={props.hideCarousel}
      className="carousel-image-modal"
      width={"100%"}
      centered
      closeIcon={
        <CloseOutlined
          data-testid="img-carousel-close-btn"
          className="close-carousel-icon"
        />
      }
      destroyOnClose={true}
      footer={null}
    >
      <div
        data-testid="offer-image-carousel-content"
        className="gallery-parent"
      >
        <Checkbox.Group value={props.selectedImages}>
          <ImageGallery
            items={imgList}
            startIndex={props.startIndex}
            showThumbnails={false}
            renderRightNav={(onClick, disabled) => {
              return (
                <Button
                  type="text"
                  icon={<RightCircleFilled className="carousel-btn" />}
                  className="image-gallery-right-nav"
                  onClick={onClick}
                ></Button>
              );
            }}
            renderLeftNav={(onClick, disabled) => {
              return (
                <Button
                  className="image-gallery-left-nav"
                  type="text"
                  icon={<LeftCircleFilled className="carousel-btn" />}
                  onClick={onClick}
                ></Button>
              );
            }}
            showPlayButton={false}
            showFullscreenButton={false}
            useBrowserFullscreen={true}
            renderItem={(image, index) => {
              return (
                <Fragment>
                  <Image
                    preview={false}
                    data-testid={image.key + "-img-carousel"}
                    className="img-carousel"
                    src={image.original}
                    onError={(ev) => (ev.target.src = defaultImg)}
                  />
                  <Row justify="center" className="chekbox-row">
                    <Checkbox
                      key={image.key}
                      className="chekbox"
                      onClick={(value) => {
                        currentImage = image.original;
                      }}
                      onChange={onChange}
                      value={image.original}
                      data-testid={
                        image.key + "-carousel-image-checkbox-display"
                      }
                    ></Checkbox>
                    <div className="img-name">
                      {getImageName(image.original) || (
                        <FormattedMessage id="nemo.unknown" />
                      )}
                    </div>
                  </Row>
                </Fragment>
              );
            }}
          />
        </Checkbox.Group>
      </div>
    </Modal>
  );
}
