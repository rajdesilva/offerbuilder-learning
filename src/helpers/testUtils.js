import React from "react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { render as rtlRender } from "@testing-library/react";
import { getValue } from "./utility";
import rootReducers from "../redux/rootReducers";
import * as languages from "../langs";
import { IntlProvider } from "react-intl";
const initialReducerState = {};
window.getValue = getValue;

window.google = {
  maps: {
    Marker: class {},
    Map: class {},
    LatLngBounds: class {},
    places: {
      Autocomplete: class {},
      AutocompleteService: class {},
      PlacesServiceStatus: {
        INVALID_REQUEST: "INVALID_REQUEST",
        NOT_FOUND: "NOT_FOUND",
        OK: "OK",
        OVER_QUERY_LIMIT: "OVER_QUERY_LIMIT",
        REQUEST_DENIED: "REQUEST_DENIED",
        UNKNOWN_ERROR: "UNKNOWN_ERROR",
        ZERO_RESULTS: "ZERO_RESULTS",
      },
      PlacesAutocomplete: {
        INVALID_REQUEST: "INVALID_REQUEST",
        NOT_FOUND: "NOT_FOUND",
        OK: "OK",
        OVER_QUERY_LIMIT: "OVER_QUERY_LIMIT",
        REQUEST_DENIED: "REQUEST_DENIED",
        UNKNOWN_ERROR: "UNKNOWN_ERROR",
        ZERO_RESULTS: "ZERO_RESULTS",
      },
    },

    MarkerClusterer: class {},
    Geocoder: class {},
  },
};
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

function render(
  ui,
  {
    initialState = initialReducerState,
    store = createStore(rootReducers, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <IntlProvider locale="en" messages={languages["en"]}>
          {children}
        </IntlProvider>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
