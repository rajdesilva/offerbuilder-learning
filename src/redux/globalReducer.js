import { globalActions } from "./globalActions";
import produce from "immer";

const initialState = {
  error: "",
  success: "",
  language: "",
};
export const globalReducer = produce((draft, action) => {
  switch (action.type) {
    case globalActions.SET_ERROR:
      draft.success = "";
      draft.error = action.payload;
      return;
    case globalActions.SET_SUCCESS:
      draft.error = "";
      draft.success = action.payload;
      return;
    case globalActions.SET_LANGUAGE:
      draft.language = localStorage.getItem('language')
      return;
    case globalActions.CLEAR:
      return initialState;
    default:
      return draft;
  }
}, initialState);
