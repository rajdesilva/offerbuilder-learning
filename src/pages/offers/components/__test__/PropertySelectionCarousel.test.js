import React from "react";
import { createStore } from "redux";
import { render, fireEvent } from "../../../../helpers/testUtils";
import { MemoryRouter } from "react-router-dom";
import PropertySelectionCarousel from "../PropertySelectionCarousel";

import { tempStore } from "./tempStore";
import { cloneDeep } from "lodash";

window.fetchWrapper = window.fetchWrapper = jest.fn().mockResolvedValue({
  success: true,
});
jest.mock("./../../service");

describe("PropertySelectionCarousel component test", () => {
  test("create PropertySelectionCarousel components snapshot", () => {
    const selectImageStore = cloneDeep(tempStore);
    selectImageStore.newoffermarketinginfo.marketingInfo.images = [
      {
        id: "Screenshot 2021-03-30 at 1-14-53 PM.png",
        uploadedBy: "swapnil.deshmukh@kognitiv.com",
        checksum: "60eef42bf8d3b2a6578f59001af2fcf7",
        url: "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png",
        createdAt: "2021-03-30T09:45:15.200651",
        updatedAt: "2021-03-30T09:45:15.200651",
        offerDetails: [],
      },
    ];
    const store = createStore(() => ({ ...selectImageStore }));
    store.dispatch = jest.fn();
    const props = {
      updateSelectedImages: jest.fn(),
      hideCarousel: jest.fn(),
      startIndex: 0,
      selectedImages: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png",
      ],
      images: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png",
      ],
    };
    const { container } = render(
      <MemoryRouter>
        <PropertySelectionCarousel {...props} />
      </MemoryRouter>,
      {
        store,
      }
    );
    expect(container).toMatchSnapshot();
  });
  test("PropertySelectionCarousel components- click on close btn and dismiss carousel", async () => {
    const selectImageStore = cloneDeep(tempStore);
    selectImageStore.newoffermarketinginfo.marketingInfo.images = [
      {
        id: "Screenshot 2021-03-30 at 1-14-53 PM.png",
        uploadedBy: "swapnil.deshmukh@kognitiv.com",
        checksum: "60eef42bf8d3b2a6578f59001af2fcf7",
        url: "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png",
        createdAt: "2021-03-30T09:45:15.200651",
        updatedAt: "2021-03-30T09:45:15.200651",
        offerDetails: [],
      },
    ];
    const store = createStore(() => ({ ...selectImageStore }));
    store.dispatch = jest.fn();
    const props = {
      updateSelectedImages: jest.fn(),
      hideCarousel: jest.fn(),
      startIndex: 0,
      selectedImages: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png",
      ],
      images: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png",
      ],
    };
    const { getByTestId } = render(
      <MemoryRouter>
        <PropertySelectionCarousel {...props} />
      </MemoryRouter>,
      {
        store,
      }
    );
    const cancelBtn = getByTestId("img-carousel-close-btn");
    fireEvent.click(cancelBtn);
    expect(props.hideCarousel).toHaveBeenCalledTimes(1);
  });

  test("PropertySelectionCarousel components- image with checkbox is unchecked if image not exists", async () => {
    const selectImageStore = cloneDeep(tempStore);
    selectImageStore.newoffermarketinginfo.marketingInfo.images = [];
    const store = createStore(() => ({ ...selectImageStore }));
    store.dispatch = jest.fn();
    const props = {
      updateSelectedImages: jest.fn(),
      hideCarousel: jest.fn(),
      startIndex: 0,
      selectedImages: [],
      images: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png",
      ],
    };
    const { getByTestId } = render(
      <MemoryRouter>
        <PropertySelectionCarousel {...props} />
      </MemoryRouter>,
      {
        store,
      }
    );
    expect(getByTestId("0-carousel-image-checkbox-display")).toBeTruthy();
    const imgCheckbox = getByTestId("0-carousel-image-checkbox-display");
    expect(imgCheckbox).toBeTruthy();
    expect(imgCheckbox.value).toBe(
      "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png"
    );
    expect(imgCheckbox.checked).toEqual(false);
  });

  test("PropertySelectionCarousel components- image with checkbox is checked if image already exists", async () => {
    const selectImageStore = cloneDeep(tempStore);
    selectImageStore.newoffermarketinginfo.marketingInfo.images = [];
    const store = createStore(() => ({ ...selectImageStore }));
    store.dispatch = jest.fn();
    const props = {
      updateSelectedImages: jest.fn(),
      hideCarousel: jest.fn(),
      startIndex: 0,
      selectedImages: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png",
      ],
      images: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png",
      ],
    };
    const { getByTestId } = render(
      <MemoryRouter>
        <PropertySelectionCarousel {...props} />
      </MemoryRouter>,
      {
        store,
      }
    );
    expect(getByTestId("0-carousel-image-checkbox-display")).toBeTruthy();
    const imgCheckbox = getByTestId("0-carousel-image-checkbox-display");
    expect(imgCheckbox).toBeTruthy();
    expect(imgCheckbox.value).toBe(
      "https://res.cloudinary.com/seekda-dev/image/upload/v1617090314/offer_builder_dev/Screenshot%202021-03-30%20at%201-14-53%20PM.png"
    );
    expect(imgCheckbox.checked).toEqual(true);
  });

  test("PropertySelectionCarousel components- if images are empty then carousel is hidden and not displayed unless images is not empty", async () => {
    const selectImageStore = cloneDeep(tempStore);
    selectImageStore.newoffermarketinginfo.marketingInfo.images = [];
    const store = createStore(() => ({ ...selectImageStore }));
    store.dispatch = jest.fn();
    const props = {
      updateSelectedImages: jest.fn(),
      hideCarousel: jest.fn(),
      startIndex: 0,
      selectedImages: [],
      images: [],
    };
    const { queryByTestId } = render(
      <MemoryRouter>
        <PropertySelectionCarousel {...props} />
      </MemoryRouter>,
      {
        store,
      }
    );
    expect(queryByTestId("0-img-carousel")).not.toBeInTheDocument();
    expect(props.hideCarousel).toHaveBeenCalledTimes(1);
  });
});
