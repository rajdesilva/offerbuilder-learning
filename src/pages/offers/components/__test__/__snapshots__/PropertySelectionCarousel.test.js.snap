// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertySelectionCarousel component test create PropertySelectionCarousel components snapshot 1`] = `<div />`;
